build and inspect

japaric opened the pull request:
build and inspect

this commits extends the UI with the following flag:

- `--bin`
- `--example`
- `--lib`
- `--release`

if one of `--bin`, `--example` or `--lib` is used then cargo-$tool will first
build the Cargo project and then proceed to run the llvm tool on the output
artifact -- that is the output artifact will be the first argument passed to the
llvm tool.

`--release` controls whether the project is compiled with optimizations or not.

if none of these three flags is passed then the old behavior is used. No path to
an artifact is passed to the llvm tool and the user is expected to provide it.

this new logic supports workspaces and the CARGO_TARGET_DIR env variable.

closes #15

r? @rust-embedded/tools (anyone)
This builds on top on #19 so that should be merged first

---

### Sample outputs:

``` rust
$ cargo new --lib foo && cd $_

$ cat >src/lib.rs <<'EOF'
#![no_std]
#[no_mangle]
pub fn foo(x: u32, y: u32) -> u32 {
    x + y
}
EOF

$ cargo objdump --lib --release -- -disassemble 2>/dev/null

/home/japaric/tmp/foo/target/release/libfoo.rlib(foo-136cd7af94a84b98.foo.awvzjwiu-cgu.0.rcgu.o):       file format ELF64-x86-64

Disassembly of section .text.foo:
foo:
       0:       8d 04 37        leal    (%rdi,%rsi), %eax
       3:       c3      retq

$ cargo objdump --target aarch64-unknown-linux-gnu --lib --release -- -disassemble 2>/dev/null

/home/japaric/tmp/foo/target/aarch64-unknown-linux-gnu/release/libfoo.rlib(foo-05ebc283581f4ba3.foo.9kf7hsz7-cgu.0.rcgu.o):     file format ELF64-aarch64-little

Disassembly of section .text.foo:
foo:
       0:       20 00 00 0b     add     w0, w1, w0
       4:       c0 03 5f d6     ret

$ cargo objdump --target riscv32imac-unknown-none-elf --lib --release -- -disassemble 2>/dev/null

/home/japaric/tmp/foo/target/riscv32imac-unknown-none-elf/release/libfoo.rlib(foo-05ebc283581f4ba3.foo.9kf7hsz7-cgu.0.rcgu.o):  file format ELF32-riscv

Disassembly of section .text.foo:
foo:
       0:       33 85 a5 00     add     a0, a1, a0
       4:       67 80 00 00     ret

$ cargo objdump --target thumbv7m-none-eabi --lib --release -- -disassemble 2>/dev/null

/home/japaric/tmp/foo/target/thumbv7m-none-eabi/release/libfoo.rlib(foo-05ebc283581f4ba3.foo.9kf7hsz7-cgu.0.rcgu.o):    file format ELF32-arm-little

Disassembly of section .text.foo:
foo:
       0:       08 44   add     r0, r1
       2:       70 47   bx      lr
```

Help text:

``` console
$ cargo nm -h
cargo-nm 0.1.2
Proxy for the `llvm-nm` tool shipped with the Rust toolchain.

USAGE:
    cargo-nm [FLAGS] [OPTIONS] [--] [args]...

FLAGS:
    -h, --help       Prints help information
        --lib        Build only this package's library
        --release    Build artifacts in release mode, with optimizations
    -V, --version    Prints version information
    -v, --verbose    Use verbose output

OPTIONS:
        --bin <NAME>         Build only the specified binary
        --example <NAME>     Build only the specified example
        --target <TRIPLE>    Target triple for which the code is compiled

ARGS:
    <args>...

The specified <args>... will all be passed to the final tool invocation.
```

japaric commented:
ping @ryankurte @therealprof anyone of you up for reviewing this?

therealprof commented on src/cargo.rs:
```diff
@@ -0,0 +1,161 @@
+use {util, Artifact, Result};
+
+pub fn artifact(
+    kind: Artifact,
```
clippy suggests `kind: &Artifact`

therealprof commented on src/cargo.rs:
```diff
@@ -0,0 +1,161 @@
+//
+// This also return the name of the crate
+pub fn target_dir() -> Result<(PathBuf, String)> {
+    let cwd = PathBuf::from(env::current_dir()?);
```
clippy suggests: `let cwd = env::current_dir()?;`

therealprof commented on src/lib.rs:
```diff
@@ -114,6 +102,12 @@ impl Context {
+            .unwrap_or(host);
+        let cfg = rustc::Cfg::parse(&target)?;
+
         for entry in WalkDir::new(sysroot.trim()).into_iter() {
```
clippy suggests: `for entry in WalkDir::new(sysroot.trim()) {`

therealprof commented on src/lib.rs:
```diff
@@ -137,117 +129,245 @@ impl Context {
-    let stderr = io::stderr();
-    let mut stderr = stderr.lock();
+    if let Some(kind) = artifact {
+        let artifact = cargo::artifact(kind, release, target_flag, ctxt.build_target())?;
```
Would need to be: `let artifact = cargo::artifact(&kind, release, target_flag, ctxt.build_target())?;` to match up with clippy suggestion

therealprof commented:
Other than the ergonomics, LGTM.

japaric commented:
I fixed the clippy warnings and added some examples to the help text. It now looks like this:

``` console
$ cargo nm -h
cargo-nm 0.1.2
Proxy for the `llvm-nm` tool shipped with the Rust toolchain.

USAGE:
    cargo-nm [FLAGS] [OPTIONS] [--] [args]...

FLAGS:
    -h, --help       Prints help information
        --lib        Build only this package's library
        --release    Build artifacts in release mode, with optimizations
    -V, --version    Prints version information
    -v, --verbose    Use verbose output

OPTIONS:
        --bin <NAME>         Build only the specified binary
        --example <NAME>     Build only the specified example
        --target <TRIPLE>    Target triple for which the code is compiled

ARGS:
    <args>...

The specified <args>... will all be passed to the final tool invocation.

To see all the flags the proxied tool accepts run `cargo-nm -- -help`.

EXAMPLES

`cargo nm --lib`                           - lists all symbols
`cargo nm --lib -- -print-size -size-sort` - lists all symbols sorted by size (smallest first)
```

I also fixed the postprocessing logic applied to the output of `size`; it was postprocessing the output in Berkely format when it shouldn't.

therealprof commented:
The behaviour is still very weird, if I specify an unexpected parameter I get a useless message:
```
error: Found argument '--disassemble' which wasn't expected, or isn't valid in this context

USAGE:
    cargo-objdump <binary-name> --release

For more information try --help
```

But if I use `--help` I get a useful message like the one posted above. Why can't I get the same useful message in both cases?

japaric commented:
Fixed the clippy warning and tweaked the help text to make clearer that the arguments for the tool must be passed after the `--`.

> Why can't I get the same useful message in both cases?

That's how clap behaves. I'm not sure if there's a way to improve the error messages. You get similar error messages if you misuse cargo-rustc:

``` console
$ cargo rustc --release  --example i2c_hal_ssd1306alphabeter --emit=obj
cargo rustc --release  --example i2c_hal_ssd1306alphabeter --emit=obj
error: Found argument '--emit' which wasn't expected, or isn't valid in this context

USAGE:
    cargo rustc --example <NAME>... --release

For more information try --help
```

therealprof commented:
Hm, we should take it up with clap then. It's certainly a usability deficiency.

therealprof commented:
bors r+
